File: ui/src/lib/match_replace.ts
// Match & replace engine — JavaScript mirror of
// `crates/bk-proxy/src/match_replace.rs`.
//
// Used by the UI's `MatchReplaceEditor` "Test" button
// (Phase 7 C-B.2) to preview what the M&R rules would do
// to a sample URL before the user clicks "Add". Runs
// client-side — no Tauri round-trip needed for a UI
// affordance.
//
// **Scope for v1:** the JS engine handles `RequestUrl`
// target only (the most common case for the "Test" button).
// `RequestHeader` and `RequestBody` targets are stubs that
// return the input unchanged — the v0.5+ "Test on a full
// request" follow-up would extend this. The Rust engine
// is the source of truth for the wire; the JS engine is a
// best-effort UI preview. If the user has a rule that
// transforms a header, the "Test" button won't show it —
// the user has to send a real replay to see the full
// effect.

import type { MatchReplaceRule } from "../types/domain";

/**
 * Apply a list of M&R rules to a sample URL string. Returns
 * the URL after all enabled rules have been applied in
 * descending-priority order. The original `url` is not
 * mutated.
 *
 * The algorithm mirrors the Rust engine's `MatchReplace::apply`:
 * 1. Filter to enabled rules.
 * 2. Sort by descending priority (high priority runs first).
 * 3. Apply each rule via `rewriteString` in turn.
 * 4. If the rewritten URL doesn't parse as a valid URL,
 *    return the previous value (the engine's defensive
 *    fallback — a bad rule shouldn't crash the preview).
 */
export function matchReplaceApplyUrl(
  url: string,
  rules: MatchReplaceRule[],
): string {
  const enabled = rules.filter((r) => r.enabled);
  // Sort by descending priority. Stable sort preserves
  // user-declared order for equal-priority rules (same as
  // the Rust `sort_by_key` with `Reverse`).
  const sorted = [...enabled].sort((a, b) => b.priority - a.priority);

  let current = url;
  for (const rule of sorted) {
    if (rule.target !== "request_url") {
      // v1: skip header/body targets. The preview is
      // URL-only. A header-body preview is a v0.5+
      // follow-up.
      continue;
    }
    current = rewriteString(current, rule);
  }
  return current;
}

/**
 * Apply a single rule to a string. Literal-string replace
 * for `is_regex: false`, regex `String.replace` for true.
 * `case_insensitive: true` adds the `i` flag.
 *
 * The Rust engine uses the `regex` crate with `(?i)` inline
 * for case-insensitive. JS's `RegExp` uses the `i` flag
 * directly. The semantics match: case-insensitive matching
 * is opt-in per rule.
 */
function rewriteString(input: string, rule: MatchReplaceRule): string {
  if (rule.is_regex) {
    try {
      const flags = rule.case_insensitive ? "gi" : "g";
      const re = new RegExp(rule.pattern, flags);
      // `replaceAll`-equivalent via the `g` flag.
      return input.replace(re, rule.replace);
    } catch {
      // Malformed regex — return input unchanged. The
      // engine's `try/catch` is the same defensive shape
      // as the Rust engine's `regex::Regex::new(...).ok()`
      // pattern: a single bad user rule shouldn't crash
      // the whole preview.
      return input;
    }
  }
  // Literal string replace. `String.prototype.replaceAll` is
  // available in all modern browsers (ES2021) and in the
  // v0.5+ Node 20 runtime.
  if (rule.case_insensitive) {
    // Build a case-insensitive literal replace.
    // `String.replaceAll` with a string arg is case-sensitive;
    // use a regex with the `i` flag for case-insensitive
    // literal replace.
    const escaped = escapeRegExp(rule.pattern);
    const re = new RegExp(escaped, "gi");
    return input.replace(re, rule.replace);
  }
  return input.split(rule.pattern).join(rule.replace);
}

/**
 * Escape a string for use as a literal in a `RegExp`
 * constructor. Mirrors the `regex::escape` function
 * from Rust's `regex` crate. Required for the
 * case-insensitive literal branch above.
 */
function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

File: crates/bk-proxy/src/match_replace.rs
//! Match & replace engine. Applies a list of `MatchReplaceRule`s to a
//! `bk_core::Request` in descending-priority order. Disabled rules are
//! skipped. The engine is a pure function — no I/O, no async — so the
//! UI's editor can preview the rewrites in real time without a Tauri
//! round-trip.
//!
//! Phase 6 (§6.3). See the §6.3 design spec for the priority + enabled
//! semantics; this file is the implementation.
//!
//! **Drift from the spec:** the spec's §6.3 used
//! `regex::RegexBuilder::new(&rule.pattern).case_insensitive(rule.case_insensitive).build()`;
//! the v0.5 `bk_core::scope` shim uses the simpler
//! `regex::Regex::new(&rule.pattern)` (with inline `(?i)` flag for
//! case-insensitive) — and the 1.10 workspace `regex` crate supports
//! both. We use the simpler form here for symmetry with the
//! `bk_core::ScopeRule::matches` `PathRegex` branch.

use bk_core::scope::{MatchReplaceRule, MatchReplaceTarget};
use bk_core::{Body, HeaderMap, Request};
use bytes::Bytes;

pub struct MatchReplace;

impl MatchReplace {
    /// Apply the rules to a request, returning a new (or
    /// mutated) `Request`. The rules are sorted by descending
    /// priority once at the top; higher priority runs first, so
    /// a high-priority rule can rewrite the URL/path/header that
    /// a lower-priority rule would otherwise have matched.
    ///
    /// **Mutation is in place** — `Request` is `Clone`, the
    /// caller is expected to pass a fresh `Request` (or the
    /// spec's documented round-trip copy from `http::Request`).
    pub fn apply(mut request: Request, rules: &[MatchReplaceRule]) -> Request {
        // Filter to enabled rules, then sort by descending priority.
        // `sort_by_key` with `Reverse` is the idiomatic way to get
        // a stable descending sort on a `Copy` key.
        let mut sorted: Vec<&MatchReplaceRule> = rules.iter().filter(|r| r.enabled).collect();
        sorted.sort_by_key(|r| std::cmp::Reverse(r.priority));

        for rule in sorted {
            request = Self::apply_one(request, rule);
        }
        request
    }

    /// Apply a single rule to the request. Routes by `target`
    /// (which part of the message to operate on). Response
    /// targets are a no-op for a `Request`-typed input — the
    /// engine just returns the request unchanged.
    fn apply_one(mut request: Request, rule: &MatchReplaceRule) -> Request {
        match rule.target {
            MatchReplaceTarget::RequestUrl => {
                let url_str = request.url.as_str().to_string();
                let new_url_str = Self::rewrite_string(&url_str, rule);
                if let Ok(parsed) = url::Url::parse(&new_url_str) {
                    // `bk_core::Url` is `url::Url` re-exported; the
                    // conversion is identity. We swallow the
                    // (unreachable) `Result` because the spec's
                    // §6.3 "Heads up" line 431 called this out
                    // as overly defensive — but `Url::parse`
                    // is fallible and a bad rewrite (e.g. an
                    // M&R rule that produces a malformed URL)
                    // should NOT crash the proxy, so the
                    // `if let Ok` is the right gate.
                    request.url = parsed;
                }
                request
            }
            MatchReplaceTarget::RequestHeader => {
                let mut new_headers: Vec<(String, String)> = Vec::new();
                for (k, v) in &request.headers {
                    let key_str = k.as_str().to_string();
                    // `HeaderValue::to_str` returns `Result<&str, ToStrError>`;
                    // a non-ASCII header value can't be string-rewritten,
                    // so we pass it through unchanged (a literal match
                    // against the raw bytes would not be useful in any
                    // case — header values are usually short identifiers
                    // and tokens).
                    let original = v.to_str().unwrap_or("").to_string();
                    let new_value = Self::rewrite_string(&original, rule);
                    new_headers.push((key_str, new_value));
                }
                let mut header_map = HeaderMap::new();
                for (k, v) in new_headers {
                    if let Ok(name) = http::HeaderName::from_bytes(k.as_bytes()) {
                        if let Ok(val) = http::HeaderValue::from_str(&v) {
                            header_map.insert(name, val);
                        }
                    }
                }
                request.headers = header_map;
                request
            }
            MatchReplaceTarget::RequestBody => {
                if let Body::Complete { data } = &request.body {
                    let s = String::from_utf8_lossy(data);
                    let new_s = Self::rewrite_string(&s, rule);
                    request.body = Body::Complete {
                        data: Bytes::from(new_s.into_bytes()),
                    };
                }
                request
            }
            // `MatchReplaceTarget` is `#[non_exhaustive]` (Phase 10
            // contract); wildcard for forward-compat. Unknown
            // targets in v1 are a no-op (the engine just passes the
            // request through).
            MatchReplaceTarget::ResponseHeader | MatchReplaceTarget::ResponseBody | _ => request,
        }
    }

    /// Apply a single rule's pattern+replace to one string.
    /// Bad regexes return the input unchanged (defense in
    /// depth: a single bad user rule should not crash the
    /// proxy).
    fn rewrite_string(input: &str, rule: &MatchReplaceRule) -> String {
        if rule.is_regex {
            // Add the `(?i)` flag inline for case-insensitive
            // matches — keeps the API surface narrow
            // (`regex::Regex::new` only, no `RegexBuilder`).
            let pat = if rule.case_insensitive {
                format!("(?i){}", rule.pattern)
            } else {
                rule.pattern.clone()
            };
            match regex::Regex::new(&pat) {
                Ok(re) => re.replace_all(input, rule.replace.as_str()).to_string(),
                Err(_) => input.to_string(), // bad regex: no change
            }
        } else {
            // Literal string replace.
            if rule.case_insensitive {
                let pat = rule.pattern.to_lowercase();
                let mut result = input.to_string();
                let mut start = 0;
                while let Some(pos) = result[start..].to_lowercase().find(&pat) {
                    let abs = start + pos;
                    let replace_len = rule.pattern.len();
                    let replacement_len = rule.replace.len();
                    result = format!(
                        "{}{}{}",
                        &result[..abs],
                        rule.replace,
                        &result[abs + replace_len..]
                    );
                    start = abs + replacement_len;
                }
                result
            } else {
                input.replace(&rule.pattern, &rule.replace)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use bk_core::{Method, Version};

    fn literal(target: MatchReplaceTarget, pattern: &str, replace: &str) -> MatchReplaceRule {
        MatchReplaceRule {
            target,
            case_insensitive: false,
            is_regex: false,
            pattern: pattern.to_string(),
            replace: replace.to_string(),
            enabled: true,
            priority: 0,
        }
    }

    fn make_get_request(url_str: &str) -> Request {
        Request {
            method: Method::GET,
            url: url_str.parse().unwrap(),
            version: Version::HTTP_11,
            headers: HeaderMap::new(),
            body: Body::empty(),
        }
    }

    #[test]
    fn literal_string_replace_in_url() {
        let req = make_get_request("https://acme.bb/api/v1/users");
        let rules = vec![literal(
            MatchReplaceTarget::RequestUrl,
            "/api/v1/",
            "/api/v2/",
        )];
        let out = MatchReplace::apply(req, &rules);
        assert_eq!(out.url.path(), "/api/v2/users");
    }

    #[test]
    fn no_matching_rule_is_identity() {
        let req = make_get_request("https://acme.bb/api/users");
        let rules = vec![literal(
            MatchReplaceTarget::RequestUrl,
            "/admin",
            "/blocked",
        )];
        let out = MatchReplace::apply(req.clone(), &rules);
        assert_eq!(out.url.path(), req.url.path());
    }

    #[test]
    fn higher_priority_runs_first() {
        let req = make_get_request("https://acme.bb/api/v1/foo");
        let mut rules = vec![
            literal(
                MatchReplaceTarget::RequestUrl,
                "/api/v1/",
                "/replaced-by-1/",
            ),
            literal(
                MatchReplaceTarget::RequestUrl,
                "/replaced-by-1/",
                "/replaced-by-2/",
            ),
        ];
        // Set the second rule to higher priority so it runs first.
        rules[0].priority = 0;
        rules[1].priority = 10;
        let out = MatchReplace::apply(req, &rules);
        // The first rule (priority 0) should NOT have replaced because
        // the second (priority 10) ran first and changed the URL.
        assert_eq!(out.url.path(), "/replaced-by-1/foo");
    }

    #[test]
    fn disabled_rules_are_skipped() {
        let req = make_get_request("https://acme.bb/api/v1/foo");
        let mut rule = literal(MatchReplaceTarget::RequestUrl, "/api/v1/", "/replaced/");
        rule.enabled = false;
        let rules = vec![rule];
        let out = MatchReplace::apply(req.clone(), &rules);
        assert_eq!(out.url.path(), req.url.path());
    }
}
